### PR TITLE
refactor NotIterator - [MOD-9677]

### PR DESCRIPTION
**Pull request by JoanFM**
## Describe the changes in the pull request

This will be the continuation of #6307 

The PR does:

- [x] Provide the functionality of the **NotIterator**
- [x] Adds testing
- [x] Adds micro-benchmarks

#### Mark if applicable

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

## Note 

I strongly believe that the OldIterator has a bug. 

The Old implementation assumes that there are no DocID in the `childIterator` that does not exist in the `wildcard iterator`(the child iterator (the negation functions) seems to be assumed to be a subgroup of the WildCard iterator). 

This test fails for OldIterator:

```cpp
#include "gtest/gtest.h"
#include "redisearch.h"
#include "index.h"
#include "query.h"
#include "micro-benchmarks/deprecated_iterator_util.h"
#include <vector>
#include <algorithm>
#include <iostream>

class NotIteratorOldTest : public ::testing::Test {
protected:
  IndexIterator *iterator_base;
  std::vector<t_docId> childDocIds = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  std::vector<t_docId> wcDocIds = {3, 4, 9};
  std::vector<t_docId> resultSet;
  t_docId maxDocId = 10;

  void SetUp() override {
    // Create the child iterator with the specified document IDs
    IndexIterator *child = (IndexIterator *)new MockOldIterator(childDocIds);

    // Create the wildcard iterator with the specified document IDs
    IndexIterator *wcii = (IndexIterator *)new MockOldIterator(wcDocIds);

    // Set up the timeout
    struct timespec timeout = {LONG_MAX, 999999999}; // "infinite" timeout

    // Create the NOT iterator with the child and wildcard iterators
    iterator_base = _New_NotIterator_With_WildCardIterator(child, wcii, maxDocId, 1.0, timeout);

    // Compute the expected result set
    // For the optimized NOT iterator, the result set is all documents in the wildcard iterator
    // that are not in the child iterator
    resultSet.clear();
    for (auto wcId : wcDocIds) {
      if (std::find(childDocIds.begin(), childDocIds.end(), wcId) == childDocIds.end()) {
        resultSet.push_back(wcId);
      }
    }

    // Print test setup information
    std::cout << "Child Doc IDs: ";
    for (auto id : childDocIds) {
      std::cout << id << " ";
    }
    std::cout << std::endl;

    std::cout << "Wildcard Doc IDs: ";
    for (auto id : wcDocIds) {
      std::cout << id << " ";
    }
    std::cout << std::endl;

    std::cout << "Expected Result Set: ";
    if (resultSet.empty()) {
      std::cout << "(empty)";
    } else {
      for (auto id : resultSet) {
        std::cout << id << " ";
      }
    }
    std::cout << std::endl;
  }

  void TearDown() override {
    iterator_base->Free(iterator_base);
  }
};

TEST_F(NotIteratorOldTest, ReadOptimized) {
  RSIndexResult *hit = iterator_base->current;
  int rc;

  std::cout << "=== READING RESULTS ===\n";
  size_t count = 0;

  // Read all results from the iterator
  while ((rc = iterator_base->Read(iterator_base->ctx, &hit)) == INDEXREAD_OK) {
    std::cout << "Read result: docId=" << hit->docId << std::endl;
    // Verify the result is in our expected result set
    //ASSERT_EQ(hit->docId, resultSet[count]);
    count++;
  }
  ASSERT_EQ(count, 0);

  // Verify we reached EOF
  ASSERT_EQ(rc, INDEXREAD_EOF);

  // Verify we read the expected number of results
  ASSERT_EQ(count, resultSet.size()) << "Expected to read " << resultSet.size() << " documents";
  std::cout << "Read " << count << " results (expected " << resultSet.size() << ")" << std::endl;

  // Try reading again after EOF
  rc = iterator_base->Read(iterator_base->ctx, &hit);
  ASSERT_EQ(rc, INDEXREAD_EOF);
  std::cout << "=== FINISHED READING ===\n";
}
```

This seems to be OK for most of the times because the QueryEvalNode filters the Not Iterator conditions before creating the ChildIterator, but there may be some edge cases where this is not true.

This makes the Benchmarks look bad for this NEW implementation.

However, I added a set of tests where the conditions are met (_Subset) in the micro benchmarks, that show that the performance is comparable or slightly better for the new implementation

### CODE REVIEW

**Comment by JoanFM** [comment]
Putting it as draft until #6240 and #6241 are merged so that this PR is smaller

File: src/iterators/not_iterator.c [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  if (ni->wcii) {
+    ni->wcii->Free(ni->wcii);
+  }
+  IndexResult_Free(base->current);
+  rm_free(base);
+}
+
+static size_t NI_NumEstimated(QueryIterator *base) {
+  NotIterator *ni = (NotIterator *)base;
+  return ni->maxDocId - ni->child->NumEstimated(ni->child);
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Thinking about it, this may lead to underestimating, while we usually expect the estimation to overestimate. Maybe we'd better keep the estimation as it is (don't consult with the child)

Comment (GuyAv46) [reply src/iterators/not_iterator.c]: also lets test this funcion

File: src/iterators/not_iterator.c [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+static IteratorStatus NI_ReadSorted_O(QueryIterator *base); // forward decl
+static IteratorStatus NI_ReadSorted_NO(QueryIterator *base); // forward decl
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Let's rename all these functions.
- No need to specify "sorted"
- Better names for optimized non-optimized implementations, instead of `_NO` sufixes

File: src/iterators/not_iterator.c [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+/* SkipTo for NOT iterator - Non-optimized version. If we have a match - return
+ * NOTFOUND. If we don't or we're at the end - return OK */
+static IteratorStatus NI_SkipTo_NO(QueryIterator *base, t_docId docId) {
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Let's add an assertion `RS_ASSERT(base->lastDocId < docId)`

File: src/iterators/not_iterator.c [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+  NotIterator *ni = (NotIterator *)base;
+  // do not skip beyond max doc id
+  if (docId > ni->maxDocId) {
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: No early return on EOF?

File: src/iterators/not_iterator.c [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    // read the next entry from the child
+    IteratorStatus rc = ni->child->SkipTo(ni->child, docId);
+    if (rc == ITERATOR_TIMEOUT) return ITERATOR_TIMEOUT;
+    if (rc != ITERATOR_OK) {
+      base->lastDocId = base->current->docId = docId;
+      return ITERATOR_OK;
+    }
+  }
+  // If the child docId is the one we are looking for, it's an anti match!
+  // We need to return NOTFOUND and set hit to the next valid docId
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]:
```suggestion
  // We need to return NOTFOUND and set the current result to the next valid docId
```

File: src/iterators/not_iterator.c [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+  IteratorStatus rc = NI_ReadSorted_NO(base);
+  if (rc == ITERATOR_OK) {
+    return ITERATOR_NOTFOUND;
+  }
+  return rc;
+}
+
+/* SkipTo for NOT iterator - Optimized version. If we have a match - return
+ * NOTFOUND. If we don't or we're at the end - return OK */
+static IteratorStatus NI_SkipTo_O(QueryIterator *base, t_docId docId) {
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Let's add an assertion `RS_ASSERT(base->lastDocId < docId)`

File: src/iterators/not_iterator.c [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+  rc = NI_ReadSorted_O(base);
+  if (rc == ITERATOR_OK) {
+    return ITERATOR_NOTFOUND;
+  }
+  return rc;
+}
+
+/* Read from a NOT iterator - Non-Optimized version. This is applicable only if
+ * the only or leftmost node of a query is a NOT node. We simply read until max
+ * docId, skipping docIds that exist in the child */
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Incorrect, let's fix this outdated comment
```suggestion
/* Read from a NOT iterator - Non-Optimized version. We simply read until max
 * docId, skipping docIds that exist in the child */
```

File: src/iterators/not_iterator.c [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  IteratorStatus rc;
+
+  if (ni->child->lastDocId == 0) {
+    // Read at least the first entry of the child
+    rc = ni->child->Read(ni->child);
+    if (rc == ITERATOR_TIMEOUT) return ITERATOR_TIMEOUT;
+  }
+  //Advance and see if it si valid
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]:
```suggestion
  // Advance and see if it is valid
```

File: src/iterators/not_iterator.c [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    rc = ni->child->Read(ni->child);
+    if (rc == ITERATOR_TIMEOUT) return rc;
+    // Check for timeout with low granularity (MOD-5512)
+    //TODO(Joan): Ticket about magic number? Should it be a config?
+    if (TimedOut_WithCtx_Gran(&ni->timeoutCtx, 5000)) {
+      base->atEOF = true;
+      return ITERATOR_TIMEOUT;
+    }
+    // The lastDocID that we proposed is not valid, so we need to try another one
+    rc = ni->wcii->Read(ni->wcii);
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Unlike the non-optimized implementation, the wildcard iterator can surpass the child iterator here. We have to address that, otherwise we may miss relevant results

File: src/iterators/not_iterator.h [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+extern "C" {
+#endif
+
+typedef struct {
+  QueryIterator base;         // base index iterator
+  QueryIterator *wcii;        // wildcard index iterator
+  QueryIterator *child;       // child index iterator
+  t_docId maxDocId;
+  TimeoutCtx timeoutCtx;
+} NotIterator, NotContext;
```
Comment (GuyAv46) [inline src/iterators/not_iterator.h]:
```suggestion
} NotIterator;
```

File: src/iterators/not_iterator.h [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+* @param weight - the weight of the node (assigned to the returned result)
+* @param timeout - the timeout for the iterator
+* @param q - the query context
+*/
+QueryIterator *IT_V2(NewNotIterator)(QueryIterator *it, t_docId maxDocId, double weight, struct timespec timeout, QueryEvalCtx *q);
+
+#ifdef __cplusplus
+}
+#endif
+
```
Comment (GuyAv46) [inline src/iterators/not_iterator.h]: Clean whitespace

File: src/iterators/not_iterator.c (hunk 9 above)
Comment (JoanFM) [reply src/iterators/not_iterator.c]: Yes, it needs better logic and better testing. Will put as draft for now

File: tests/cpptests/micro-benchmarks/benchmark_not_iterator.cpp [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+      this->SetName(#BaseClass "<" #a "," #b ", " #c ">/" #Method);       \
+    }                                                                     \
+                                                                          \
+   protected:                                                             \
+    void BenchmarkCase(::benchmark::State&) BENCHMARK_OVERRIDE;           \
+  };
+
+  #define BENCHMARK_TEMPLATE3_DEFINE_F(BaseClass, Method, a, b, c)    \
+  BENCHMARK_TEMPLATE3_PRIVATE_DECLARE_F(BaseClass, Method, a, b, c)   \
+  void BENCHMARK_PRIVATE_CONCAT_NAME(BaseClass, Method)::BenchmarkCase
```
Comment (GuyAv46) [inline tests/cpptests/micro-benchmarks/benchmark_not_iterator.cpp]: CAn't we use the generic `BENCHMARK_TEMPLATE_DEFINE_F` macro? We already require C++17

File: tests/cpptests/iterator_util.h [hunk 13]
```diff
@@ -31,74 +33,94 @@ class MockIterator {
...
-            base.atEOF = true;
-            return whenDone;
-        }
-        base.lastDocId = base.current->docId = docIds[nextIndex++];
-        return ITERATOR_OK;
+      if (sleepTime.has_value()) {
+        std::this_thread::sleep_for(sleepTime.value());
+      }
+      readCount++;
+      if (nextIndex >= docIds.size() || base.atEOF || docIds.size() == 0) {
```
Comment (GuyAv46) [inline tests/cpptests/iterator_util.h]: If `docIds.size() == 0` then `nextIndex >= docIds.size()`, no? Do we need this validation?

File: tests/cpptests/iterator_util.h [hunk 14]
```diff
@@ -31,74 +33,94 @@ class MockIterator {
...
+      docIds.erase(new_end, docIds.end());
+    }
+
+    template<typename... Args>
+    MockIterator(std::chrono::milliseconds sleep, Args&&... args)
+      : docIds({std::forward<Args>(args)...}), whenDone(ITERATOR_EOF), nextIndex(0), readCount(0), sleepTime(sleep) {
+      setBase(&base);
+      auto new_end = std::unique(docIds.begin(), docIds.end());
+      docIds.erase(new_end, docIds.end());
+    }
```
Comment (GuyAv46) [inline tests/cpptests/iterator_util.h]: We should stick to a single constructor and reuse it (like we had before), avoiding logic duplication

File: tests/cpptests/iterator_util.h [hunk 15]
```diff
@@ -31,74 +33,94 @@ class MockIterator {
...
 
     template<typename... Args>
-    MockIterator(IteratorStatus st, Args&&... ids_args)
-        : docIds({ids_args...}), whenDone(st), nextIndex(0), readCount(0) {
-        setBase(&base);
-        std::sort(docIds.begin(), docIds.end());
```
Comment (GuyAv46) [inline tests/cpptests/iterator_util.h]: We should keep the sorting here. Some places already rely on this

File: src/iterators/not_iterator.c [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+  return ITERATOR_EOF;
+}
+
+QueryIterator *IT_V2(NewNotIterator)(QueryIterator *it, t_docId maxDocId, double weight, struct timespec timeout, QueryEvalCtx *q) {
+  NotIterator *ni = rm_calloc(1, sizeof(*ni));
+  QueryIterator *ret = &ni->base;
+  bool optimized = q && q->sctx->spec->rule && q->sctx->spec->rule->index_all;
+  if (optimized) {
+    //TODO(Joan): Once all the WildCards are migrated, call the generic one
+    ni->wcii = IT_V2(NewWildcardIterator_NonOptimized)(q->docTable->maxDocId, q->docTable->size);
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: This API is ready

File: src/iterators/not_iterator.c [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+  if (ni->wcii) {
+    ni->wcii->Free(ni->wcii);
+  }
+  IndexResult_Free(base->current);
+  rm_free(base);
+}
+
+static size_t NI_NumEstimated(QueryIterator *base) {
+  NotIterator *ni = (NotIterator *)base;
+  return ni->maxDocId;
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: If we have a wildcard iterator, we better use its estimation. `maxDocId` can grow much higher than the actual number of documents. We also have the doc table size in the construction, we can hold it as well as a better estimation

File: tests/cpptests/micro-benchmarks/benchmark_not_iterator.cpp (hunk 12 above)
Comment (JoanFM) [reply tests/cpptests/micro-benchmarks/benchmark_not_iterator.cpp]: It does not have option to have 3 template parameters, they have for up to 2

File: tests/cpptests/iterator_util.h (hunk 15 above)
Comment (JoanFM) [reply tests/cpptests/iterator_util.h]: I think I did not realize I removed that

File: tests/cpptests/test_cpp_iterator_not.cpp [hunk 18]
```diff
@@ -547,7 +567,7 @@ INSTANTIATE_TEST_SUITE_P(
       }()
     ),
     ::testing::Values(std::nullopt),
-    ::testing::Values(false)
+    ::testing::Values(false, true)
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_iterator_not.cpp]:
```suggestion
    ::testing::Bool()
```

File: tests/cpptests/test_cpp_iterator_not.cpp [hunk 19]
```diff
@@ -449,7 +469,7 @@ INSTANTIATE_TEST_SUITE_P(
       std::vector<t_docId>{50, 60, 70, 80, 90, 100, 600, 750, 950, 1200}
     ),
     ::testing::Values(std::nullopt),
-    ::testing::Values(false)
+    ::testing::Values(false, true)
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_iterator_not.cpp]:
```suggestion
    ::testing::Bool()
```

File: tests/cpptests/test_cpp_iterator_not.cpp [hunk 20]
```diff
@@ -376,7 +394,7 @@ INSTANTIATE_TEST_SUITE_P(
       }()
     ),
     ::testing::Values(std::nullopt, 100),
-    ::testing::Values(false)
+    ::testing::Values(false, true)
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_iterator_not.cpp]:
```suggestion
    ::testing::Bool()
```

File: tests/cpptests/test_cpp_iterator_not.cpp [hunk 21]
```diff
@@ -109,6 +111,9 @@ class NotIteratorCommonTest : public ::testing::TestWithParam<std::tuple<std::ve
 };
 
 TEST_P(NotIteratorCommonTest, Read) {
+  if (optimized && opt_max_doc_id.has_value()) {
+    GTEST_SKIP() << "For optimized version the maxDocId is not necessarily respected (smaller than the max of Child or WCDocId): Is it worth adding this check? return base->lastDocId < ni->maxDocId ? ITERATOR_EOF : ITERATOR_OK; ";
+  }
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_iterator_not.cpp:116]: I agree that this scenario is not relevant.
Can we define the test combinations in a way that avoids redundant combinations? (Maybe taking the maximal id from the wildcard IDs vector if `optimized == false`?)

File: src/iterators/not_iterator.c [hunk 22]
```diff
@@ -74,6 +74,71 @@ static IteratorStatus NI_Read_NotOptimized(QueryIterator *base) {
...
+  if (ni->child->lastDocId == 0) {
+    // Read at least the first entry of the child
+    rc = ni->child->Read(ni->child);
+    if (rc == ITERATOR_TIMEOUT) return ITERATOR_TIMEOUT;
+  }
+
+  // Advance to the next potential docId
+  rc = ni->wcii->Read(ni->wcii);
+  if (rc == ITERATOR_TIMEOUT) {
+    base->atEOF = true;
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: we shouldn't mark the iterator as EOF on timeout

File: src/iterators/not_iterator.c [hunk 23]
```diff
@@ -74,6 +74,71 @@ static IteratorStatus NI_Read_NotOptimized(QueryIterator *base) {
...
+    // Case 3: Current docID is ahead of child's docID
+    // This means we need to advance the child until it catches up, and in next loop we will see how they compare
+    else { //(ni->child->lastDocId < ni->wcii->lastDocId)
+      while (!ni->child->atEOF && ni->child->lastDocId < ni->wcii->lastDocId) {
+        rc = ni->child->Read(ni->child);
+        if (rc == ITERATOR_TIMEOUT) return rc;
+      }
+    }
+    if (TimedOut_WithCtx_Gran(&ni->timeoutCtx, 5000)) {
+      base->atEOF = true;
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: we shouldn't mark the iterator as EOF on timeout

File: src/iterators/not_iterator.c [hunk 24]
```diff
@@ -74,6 +74,71 @@ static IteratorStatus NI_Read_NotOptimized(QueryIterator *base) {
...
+
+    // Case 2: Current docID is equal to child's docID.
+    // If we're at the same docId as the child, we need to advance both and in next loop iteration we will see how they compare
+    // TODO: Joan here we are not doing this
+    else if (ni->wcii->lastDocId == ni->child->lastDocId) {
+      rc = ni->child->Read(ni->child);
+      if (rc == ITERATOR_TIMEOUT) return rc;
+      rc = ni->wcii->Read(ni->wcii);
+      if (rc == ITERATOR_TIMEOUT) {
+        base->atEOF = true;
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: we shouldn't mark the iterator as EOF on timeout

File: src/iterators/not_iterator.c [hunk 25]
```diff
@@ -112,14 +177,113 @@ static IteratorStatus NI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId)
...
+      // Child doesn't have this document - valid result
+      base->lastDocId = base->current->docId = ni->wcii->lastDocId;
+      return base->lastDocId == docId ? ITERATOR_OK : ITERATOR_NOTFOUND;
+    }
+  }
+
+  // Default fallback (should not reach here in normal execution)
+  // LCOV_EXCL_START
+  return ITERATOR_EOF;
+  // LCOV_EXCL_STOP
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Let's avoid using these magic comments in production code. If we shouldn't reach here, can we rewrite the logic so we don't have redundant fallbacks?

File: src/iterators/not_iterator.c [hunk 26]
```diff
@@ -112,14 +177,113 @@ static IteratorStatus NI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId)
...
+ * - ni: The NotIterator
+ * - docId: The original requested document ID
+ *
+ * Returns:
+ * - ITERATOR_OK if a valid result was found and it matches the original docId
+ * - ITERATOR_NOTFOUND if a valid result was found but it's not the original docId
+ * - ITERATOR_EOF if we've reached the end
+ * - ITERATOR_TIMEOUT if the operation timed out
+ */
+static IteratorStatus NI_HandleRelativePositions(QueryIterator *base, NotIterator *ni, t_docId docId) {
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: This function is very complex and is only used once, so I suspect it has redundancies.
The implementation in #5619 is much shorter. Is there a problem with it?

File: src/iterators/not_iterator.c (hunk 23 above)
Comment (JoanFM) [reply src/iterators/not_iterator.c]: I agree, but in this case the old implementation did. Just FYI

File: tests/cpptests/test_cpp_iterator_not.cpp (hunk 21 above)
Comment (JoanFM) [reply tests/cpptests/test_cpp_iterator_not.cpp:116]: I can do, but to me is easier to understand a clear skip message than a hard-to-follow logic in the SetUp. Your call

File: src/iterators/not_iterator.c (hunk 26 above)
Comment (JoanFM) [reply src/iterators/not_iterator.c]: It seems to not account for the situation where the subset of ChildIDs is not in the WCIds

Comment (JoanFM) [reply src/iterators/not_iterator.c]: Tests fail if changing the implementation for this:

```cpp
IteratorStatus NI_SkipTo_Optimized(QueryIterator *base, t_docId docId) {
  // do not skip beyond max doc id
  NotIterator *ni = (NotIterator *)base;
  if (docId > ni->maxDocId) {
    base->atEOF = true;
    return ITERATOR_EOF;
  }

  RSIndexResult *wr;
  IteratorStatus rc = ni->wcii->SkipTo(ni->wcii, docId);
  if (rc != ITERATOR_OK && rc != ITERATOR_OK) {
    return rc;
  } else if (rc == ITERATOR_OK) {
    // A valid wildcard result was found. Let's check if the child has it
    if (ni->child->lastDocId > docId || ni->child->atEOF) {
      // If the child is ahead of the skipto id, it means the child doesn't have this id.
      // So we are okay!
      base->lastDocId = base->current->docId = docId;
      return ITERATOR_OK;
    } else if (ni->child->lastDocId < docId) {
      // read the next entry from the child
      rc = ni->child->SkipTo(ni->child, docId);
      if (rc == ITERATOR_TIMEOUT) return ITERATOR_TIMEOUT;
      if (rc != ITERATOR_OK) {
        base->lastDocId = base->current->docId = docId;
      }
    }
  }

  // If the wildcard iterator is missing the docId, or the child iterator has it,
  // We need to return NOTFOUND and set hit to the next valid docId
  rc = NI_Read_Optimized(base);
  if (rc == ITERATOR_OK) {
    return ITERATOR_NOTFOUND;
  }
  return rc;
}
```

Comment (GuyAv46) [reply src/iterators/not_iterator.c]: I see...

File: src/iterators/not_iterator.c [hunk 27]
```diff
@@ -112,14 +174,108 @@ static IteratorStatus NI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId)
...
+    return base->lastDocId == docId ? ITERATOR_OK : ITERATOR_NOTFOUND;
+  }
+  // Case 2: Both iterators at same position
+  else if (ni->wcii->lastDocId == ni->child->lastDocId) {
+    // Both at same position - find next valid result
+    rc = NI_Read_Optimized(base);
+    if (rc == ITERATOR_OK) {
+      return ITERATOR_NOTFOUND;
+    } else if (rc == ITERATOR_EOF) {
+      base->atEOF = true;
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: The call to `NI_Read_Optimized` already set it up
```suggestion
```
(We can assert that instead)
```suggestion
    } else if (rc == ITERATOR_EOF) {
      RS_ASSERT(base->atEOF);
```

File: src/iterators/not_iterator.c [hunk 28]
```diff
@@ -112,14 +174,108 @@ static IteratorStatus NI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId)
...
+        return ITERATOR_NOTFOUND;
+      } else if (rc == ITERATOR_EOF) {
+        base->atEOF = true;
+      }
+    } else if (rc == ITERATOR_NOTFOUND || rc == ITERATOR_EOF) {
+      // Child doesn't have this document - valid result
+      base->lastDocId = base->current->docId = ni->wcii->lastDocId;
+      return base->lastDocId == docId ? ITERATOR_OK : ITERATOR_NOTFOUND;
+    }
+  }
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: At this point, `child->lastDocId >= docId` (if it was initially <, we skipped it to `docId` and now it's `child->lastDocId >= docId`)
This means that this path necessarily handles `ITERATOR_NOTFOUND` case

File: src/iterators/not_iterator.c [hunk 29]
```diff
@@ -112,14 +174,108 @@ static IteratorStatus NI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId)
...
+    rc = ni->child->SkipTo(ni->child, ni->wcii->lastDocId);
+    if (rc == ITERATOR_TIMEOUT) return ITERATOR_TIMEOUT;
+
+    if (rc == ITERATOR_OK) {
+      // Child has this document - find next valid result
+      rc = NI_Read_Optimized(base);
+      if (rc == ITERATOR_OK) {
+        return ITERATOR_NOTFOUND;
+      } else if (rc == ITERATOR_EOF) {
+        base->atEOF = true;
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Should've been handled by `NI_Read_Optimized`

File: src/iterators/not_iterator.c [hunk 30]
```diff
@@ -112,14 +174,108 @@ static IteratorStatus NI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId)
...
+      }
+    } else if (rc == ITERATOR_NOTFOUND || rc == ITERATOR_EOF) {
+      // Child doesn't have this document - valid result
+      base->lastDocId = base->current->docId = ni->wcii->lastDocId;
+      return base->lastDocId == docId ? ITERATOR_OK : ITERATOR_NOTFOUND;
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Cannot be OK
```suggestion
      return ITERATOR_NOTFOUND;
```

File: src/iterators/not_iterator.c [hunk 31]
```diff
@@ -112,14 +174,108 @@ static IteratorStatus NI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId)
...
+    if (rc == ITERATOR_OK) {
+      return ITERATOR_NOTFOUND;
+    } else if (rc == ITERATOR_EOF) {
+      base->atEOF = true;
+    }
+  }
+  // Case 3: Wildcard is ahead of child
+  else { // ni->wcii->lastDocId > ni->child->lastDocId
+    // Wildcard advanced past child - check if child has this new docID
+    rc = ni->child->SkipTo(ni->child, ni->wcii->lastDocId);
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: This flow may require calling `child->SkipTo` twice:
1. `docId > child->lastDocId` so we skip to it (from `NI_SkipTo_Optimized`)
2. We call `wcii->SkipTo` (always, at the start of the function)
3. And now here

I believe removing the first call will still work and improve the performance in this case.
I also suggest inlining this function so these observations are easier to see

File: src/iterators/not_iterator.c (hunk 28 above)
Comment (JoanFM) [reply src/iterators/not_iterator.c]: not if the ChildID is atEOF right? If I do not do this, some tests fail

Comment (JoanFM) [reply src/iterators/not_iterator.c]: What I can do is if Child is atEOF, directly call wc_>SkipTo

File: src/iterators/not_iterator.c (hunk 31 above)
Comment (JoanFM) [reply src/iterators/not_iterator.c]: Removing the first call makes SkipTo fail for "Child not Subset of WildCard case".

Comment (JoanFM) [reply src/iterators/not_iterator.c]: I will do some refactor to simplify, it can be handled in a signle fucntion

File: src/iterators/not_iterator.c [hunk 32]
```diff
@@ -74,6 +74,68 @@ static IteratorStatus NI_Read_NotOptimized(QueryIterator *base) {
...
+  }
+
+  IteratorStatus rc;
+
+  // If child has not read any element, read one
+  if (ni->child->lastDocId == 0) {
+    // Read at least the first entry of the child
+    rc = ni->child->Read(ni->child);
+    if (rc == ITERATOR_TIMEOUT) return ITERATOR_TIMEOUT;
+  }
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: This step seems redundant with your implementation. Seems like we will necessarily reach case 3 and will exit the while loop by the same condition

File: src/iterators/not_iterator.c [hunk 33]
```diff
@@ -74,6 +74,68 @@ static IteratorStatus NI_Read_NotOptimized(QueryIterator *base) {
...
+      }
+    }
+    // Case 3: Current docID is ahead of child's docID
+    // This means we need to advance the child until it catches up, and in next loop we will see how they compare
+    else { //(ni->child->lastDocId < ni->wcii->lastDocId)
+      while (!ni->child->atEOF && ni->child->lastDocId < ni->wcii->lastDocId) {
+        rc = ni->child->Read(ni->child);
+        if (rc == ITERATOR_TIMEOUT) return rc;
+      }
+    }
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Let's finalize the implementation here

- [ ] remove the `TODO`s left (address them if not done already)
- [ ] Move each case comment into its case block. At the moment, it's hard to see that the conditions are connected to each other

File: src/iterators/not_iterator.c [hunk 34]
```diff
@@ -112,14 +174,94 @@ static IteratorStatus NI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId)
...
+    // Both at same position - find next valid result
+    rc = NI_Read_Optimized(base);
+    if (rc == ITERATOR_OK) {
+      return ITERATOR_NOTFOUND;
+    } else if (rc == ITERATOR_EOF) {
+      RS_ASSERT(base->atEOF);
+    }
+  }
+  // Case 3: Wildcard is ahead of child
+  else { // ni->wcii->lastDocId > ni->child->lastDocId
```
Comment (GuyAv46) [inline src/iterators/not_iterator.c]: Same comments about the case comments and `else..if` layout

File: tests/cpptests/test_cpp_iterator_not.cpp [hunk 35]
```diff
@@ -134,6 +139,9 @@ TEST_P(NotIteratorCommonTest, SkipToEOF) {
 }
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_iterator_not.cpp:139]: Fix skip-to id (skipping backwards is undefined). This means we should also push the assertion about it to the beginning of both SkipTo implementations

**Review by GuyAv46 — APPROVED** [review]
